### DISCUSSION TITLE
Error when trying to run tests

### DISCUSSION THREAD

**Discussion by Blacedraco** [typegoose/mongodb-memory-server#868] [Q&A] [answered]
## Versions

- NodeJS: 20.12.2
- mongodb-memory-server-*: 6.0.4 
- mongodb(the binary version): 6.0.4
- mongodb(the js package): 0.0.0
- mongoose: ^8.2.4 
- system: Windows, 

package: mongo-memory-server 

## What is your question?
If i try to run tests i get the following error Message:

```txt
Determining test suites to run...Starting the MongoMemoryServer Instance failed, enable debug log for more information. Error:
 UnexpectedCloseError: Instance closed unexpectedly with code "3221225781" and signal "null"
Exit Code is large, commonly meaning that vc_redist is not installed, the latest vc_redist can be found at https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170
    at MongoInstance.closeHandler (C:...\node_modules\mongodb-memory-server-core\lib\util\MongoInstance.js:302:58)
    at ChildProcess.emit (node:events:518:28)
    at maybeClose (node:internal/child_process:1105:16)
    at Socket.<anonymous> (node:internal/child_process:457:11)
    at Socket.emit (node:events:518:28)
    at Pipe.<anonymous> (node:net:337:12)
Error: Instance closed unexpectedly with code "3221225781" and signal "null"
Exit Code is large, commonly meaning that vc_redist is not installed, the latest vc_redist can be found at https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170
    at MongoInstance.closeHandler (C:...\node_modules\mongodb-memory-server-core\lib\util\MongoInstance.js:302:58)
    at ChildProcess.emit (node:events:518:28)
    at maybeClose (node:internal/child_process:1105:16)
    at Socket.<anonymous> (node:internal/child_process:457:11)
    at Socket.emit (node:events:518:28)
    at Pipe.<anonymous> (node:net:337:12)
```

and if i try to use a specific version with the following command,

`"test": " set MONGOMS_VERSION=6.0.4 && jest --runInBand",`

i get the message:

```
Error: Download failed for url "https://fastdl.mongodb.org/windows/mongodb-windows-x86_64-6.0.4%20.zip", Details:
Status Code is 403 (MongoDB's 404)
This means that the requested version-platform combination doesn't exist
Try to use different version 'new MongoMemoryServer({ binary: { version: 'X.Y.Z' } })'
List of available versions can be found here: https://www.mongodb.com/download-center/community/releases/archive
    at RedirectableRequest.<anonymous> (C:\...\node_modules\mongodb-memory-server-core\lib\util\MongoBinaryDownload.js:272:32)
    at RedirectableRequest.emit (node:events:518:28)
    at RedirectableRequest._processResponse (C:\...\node_modules\follow-redirects\index.js:398:10)
    at RedirectableRequest._onNativeResponse (C:\...\node_modules\follow-redirects\index.js:91:12)
    at Object.onceWrapper (node:events:633:26)
    at ClientRequest.emit (node:events:518:28)
    at HTTPParser.parserOnIncomingClient (node:_http_client:698:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)
    at TLSSocket.socketOnData (node:_http_client:540:22)
    at TLSSocket.emit (node:events:518:28)
```

Neither me nor my professor have any idea how to fix it, but already tried reinstalling everything.

**Comment by hasezoey** [comment]
> mongodb-memory-server-*: 6.0.4

such a mongodb-memory-server version does not exist, the last `6.0.x` version is `6.0.2`, in any case, `6.x` is not supported anymore. currently the latest version is `9.2.0`.


> If i try to run tests i get the following error Message:
> Neither me nor my professor have any idea how to fix it, but already tried reinstalling everything.

```txt
UnexpectedCloseError: Instance closed unexpectedly with code "3221225781" and signal "null"
Exit Code is large, commonly meaning that vc_redist is not installed, the latest vc_redist can be found at https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170
```

did you already try what the error suggests and install the latest `vc_redist` package?

> "test": " set MONGOMS_VERSION=6.0.4 && jest --runInBand",
> Error: Download failed for url "https://fastdl.mongodb.org/windows/mongodb-windows-x86_64-6.0.4%20.zip", Details:

something with your `set` command is likely wrong and adds a extra space to the environment variable (resulting in `6.0.4 `), which should (to my knowledge) not happen, though in the future i guess we can try to trim the input.
i dont know if this will work on windows, but does running `MONGOMS_VERSION=6.0.4 jest --runInBand` work (instead of your `set` and `&&` command)?
if that does not work and this is the package.json anyway, try setting it under `config`, see [How to use them in the package.json](https://nodkz.github.io/mongodb-memory-server/docs/api/config-options#how-to-use-them-in-the-packagejson)

**Comment by jpilgrim** [comment]
Let me try to clarify the situation a little bit:

- We (I helped @Blacedraco) first tried to use the default version, which is 6.0.14, istn't it?
- As this version did not work, we tried different other versions, one of them was 6.0.4. So we do not insist on using 6.0.4, it was just a test.
- `mongodb-memory-server-*: 6.0.4` -- That's wrong, we actually use 9.1.8. Must have been a copy-paste-error when creating the issue.
- I would consider the trailing space in the version number a bug, but as @hasezoey already wrote, this could be easily fixed.
- AFAIK "set" is the way do set an environment variable under Windows, isn't it?
- @Blacedraco: Try to run the tests with this script line (with no space before the &&):
	`"test": " set MONGOMS_VERSION=6.0.4&& jest --runInBand"`
  Maybe you could also try older MongoDB-versions, such as 5.x. 
- As far as I understand, vc_redist has already been installed manually. But I'm no windows expert.
- We downloaded some versions manually, unzipped the download, renamed (mondgod-x64-win32-...exe) and moved the exe file to c:\Users\{user}\.cache\mondodb-binaries. However, MMS did not pick up our manually installed versions, i.e. it tried to download them again (with the space-problem in the URL). I did this trick under macos where it worked. Has the location changed? Hmm... after thinking about it, I would assume that the trailing-space problem sabotaged all our tests.

**Reply by Blacedraco** [reply]
removing the space before the && didn't change anything in the output neither did changing the specified versions in that command.
Removing the set and && gave me another errormessage as it cant read that command in that constellation

**Reply by hasezoey** [reply]
> We (I helped @Blacedraco) first tried to use the default version, which is 6.0.14, istn't it?

all the default versions can be found at [Mongodb Server Versions](https://nodkz.github.io/mongodb-memory-server/docs/guides/mongodb-server-versions)

> AFAIK "set" is the way do set an environment variable under Windows, isn't it?

i dont know, i dont use windows and didnt check for this issue (as there were seemingly more glaring issues), in linux / unix you can just prepend the environment variable if it should only be valid / set for that command.

maybe try to surround the environment variable's string with quotation marks: `MONGOMS_VERSION="6.0.4"` or the full string `"MONGOMS_VERSION=6.0.4"`.

maybe you are running into [this problem on stackoverflow](https://superuser.com/questions/57723/prevent-extra-space-when-setting-variable-on-windows-command-line).

> However, MMS did not pick up our manually installed versions, i.e. it tried to download them again (with the space-problem in the URL)
> Hmm... after thinking about it, I would assume that the trailing-space problem sabotaged all our tests.

it is likely that the full version MMS was searching for was `mongodb-x64-win32-6.0.4 .exe` as the version gets added without trimming currently (as to my knowledge trailing spaces should never be there), it also never came up as a problem.

as a side note, you can use config option [`USE_ARCHIVE_NAME_FOR_BINARY_NAME`](https://nodkz.github.io/mongodb-memory-server/docs/api/config-options#use_archive_name_for_binary_name) to use the archive name as the binary name instead (though this will make it unable to read the already existing other style binaries).

> As far as I understand, vc_redist has already been installed manually. But I'm no windows expert.

i would like to help, but the only error we get is the high exist code (`3221225781`), which on windows until now always indicated a missing install of vc_redist.

**Reply by jpilgrim** [reply]
I have to correct myself: The trailing space is a problem of the Windows cmd tool. But as mentioned, there are workarounds, e.g. by defining a variable in package.json or removing the space before the &&.

**Comment by Blacedraco** [comment, marked as answer]
Changing the version number to an actual existing one an removing the space between the number and the && made it finally work.
The Version that works looks like this : ..MONGOMS_VERSION=4.0.27&&..
thanks to all that suggested solutions and helped with that problem
